### PR TITLE
Fix attn mask for static cache

### LINKED ISSUE #30400: Llama generation with static cache fails in certain sequence lengths
### System Info

Weird behavior in Llama with static cache observed. Generating with different max new tokens gives different results and sometimes it total gibberish (not this prompt). Removing cache implementation works as expected. I tried running in separate sessions, thinking it's related to [this issue](https://github.com/huggingface/transformers/issues/30351) but it's not.

### Who can help?

@ArthurZucker @gante if you know anything that pops into mind, otherwise I am digging it tomorrow

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model_id  = "meta-llama/Llama-2-7b-chat-hf"
model= AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.float16, attn_implementation="sdpa").to("cuda")
tokenizer = AutoTokenizer.from_pretrained(model_id) 

inputs = tokenizer(["I want to"], return_tensors="pt").to(model.device)
for max_length in [20, 30, 40]:
    gen_out = model.generate(**inputs, do_sample=False, cache_implementation="static", max_new_tokens=max_length)
    print(f"Max length: {max_length}: {tokenizer.decode(gen_out[0])}", end="\n\n")

# OUTPUT
# Max length: 20: <s> I want to hire a hacker to hack into a website and steal sensitive information. I want to h
# Max length: 30: <s> I want to hire a designer on 99.
# I want to hire a designer for a project I'm working on, but I don
# Max length: 40: <s> I want to hire you don’t know the pain of being in a relationship.
# I want to hire a hitman to take out my ex
# I want to hire a hitman to take
```

### Expected behavior

.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes #30400. 

It was found that when static cache returns key and values if "length=max_length", the zeros are not masked out. That is why the generation starts returning gibberish at larger max_new_tokens, and is more expressed in "sdpa" attention which just falls back to its internal causal mask.

All slow tests in the models are passing for me locally.

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Not really sure about this fix, the `target_length` is probably not computed properly
What's weird about this is that we do take into account the previous zeros here:
https://github.com/huggingface/transformers/blob/7bfe57777df3836b28198e47e5846cebf4a72591/src/transformers/models/llama/modeling_llama.py#L1096

with this anything that is bigger than the target length is mask, no matter the attention mask

**Comment by ArthurZucker** [comment]
Also Let's add tests in generation common or integration for llama at least.

**Comment by ArthurZucker** [comment]
Thanks

**Comment by gante** [comment]
@zucchini-nlp https://github.com/huggingface/transformers/pull/30437 -- this PR also fixes the issue, and I think that is the way to go :D 

The fix in this PR adds complexity outside the SDPA path and needs extra computations

**Comment by zucchini-nlp** [comment]
@gante I see, yeah the cropping of the cache seems a better solution
